In [1]:
import numpy as np
import pandas as pd

#### **Q-1 What's the version of Pandas that you installed?**

In [2]:
pd.__version__

'1.5.3'

In [3]:
df = pd.read_csv('datasets_master_housing.csv')
df.head()

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -122.23     37.88                41.0        880.0           129.0   
1    -122.22     37.86                21.0       7099.0          1106.0   
2    -122.24     37.85                52.0       1467.0           190.0   
3    -122.25     37.85                52.0       1274.0           235.0   
4    -122.25     37.85                52.0       1627.0           280.0   

   population  households  median_income  median_house_value ocean_proximity  
0       322.0       126.0         8.3252            452600.0        NEAR BAY  
1      2401.0      1138.0         8.3014            358500.0        NEAR BAY  
2       496.0       177.0         7.2574            352100.0        NEAR BAY  
3       558.0       219.0         5.6431            341300.0        NEAR BAY  
4       565.0       259.0         3.8462            342200.0        NEAR BAY

#### **Q-2 How many columns are in the dataset?**

In [5]:
df.shape[1]

10

#### **Q-3 Which columns in the dataset have missing values?**

In [9]:
df.isnull().sum()

longitude               0
latitude                0
housing_median_age      0
total_rooms             0
total_bedrooms        207
population              0
households              0
median_income           0
median_house_value      0
ocean_proximity         0
dtype: int64

**total_bedrooms**  207 missing values

#### **Q-4 How many unique values does the `ocean_proximity` column have?**

In [10]:
df['ocean_proximity'].nunique()

5

#### **Q-5 What's the average value of the `median_house_value` for the houses located near the bay?**

In [12]:
df[df['ocean_proximity'] == 'NEAR BAY']['median_house_value'].mean()

259212.31179039303

#### **Q-6**
**1 .Calculate the average of `total_bedrooms` column in the dataset.**

In [13]:
df['total_bedrooms'].mean()

537.8705525375618

**2. Use the `fillna` method to fill the missing values in `total_bedrooms` with the mean value from the previous step.**

In [14]:
df['total_bedrooms'] = df['total_bedrooms'].fillna(df['total_bedrooms'].mean())

**3. Now, calculate the average of `total_bedrooms` again.**

In [15]:
df['total_bedrooms'].mean()

537.8705525375617

**4. Has it changed?** \
Hint: take into account only 3 digits after the decimal point.

#### **Q-7**

**1. Select all the options located on islands.**

In [21]:
df[df['ocean_proximity'] == 'ISLAND']

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
8314    -118.32     33.35                27.0       1675.0           521.0   
8315    -118.33     33.34                52.0       2359.0           591.0   
8316    -118.32     33.33                52.0       2127.0           512.0   
8317    -118.32     33.34                52.0        996.0           264.0   
8318    -118.48     33.43                29.0        716.0           214.0   

      population  households  median_income  median_house_value  \
8314       744.0       331.0         2.1579            450000.0   
8315      1100.0       431.0         2.8333            414700.0   
8316       733.0       288.0         3.3906            300000.0   
8317       341.0       160.0         2.7361            450000.0   
8318       422.0       173.0         2.6042            287500.0   

     ocean_proximity  
8314          ISLAND  
8315          ISLAND  
8316          ISLAND  
8317          ISLAND  
8318          ISLAND

**2. Select only columns `housing_median_age`, `total_rooms`, `total_bedrooms`.**

In [22]:
df[df['ocean_proximity'] == 'ISLAND'][['housing_median_age', 'total_rooms', 'total_bedrooms']]

housing_median_age  total_rooms  total_bedrooms
8314                27.0       1675.0           521.0
8315                52.0       2359.0           591.0
8316                52.0       2127.0           512.0
8317                52.0        996.0           264.0
8318                29.0        716.0           214.0

**3. Get the underlying NumPy array. Let's call it X.**

In [25]:
X = df[df['ocean_proximity'] == 'ISLAND'][['housing_median_age', 'total_rooms', 'total_bedrooms']].values

In [26]:
X

array([[  27., 1675.,  521.],
       [  52., 2359.,  591.],
       [  52., 2127.,  512.],
       [  52.,  996.,  264.],
       [  29.,  716.,  214.]])

**4. Compute matrix-matrix multiplication between the transpose of `X` and `X`. To get the transpose, use `X.T`. Let's call the result `XTX`**

In [28]:
XTX = X.T.dot(X)

In [29]:
XTX

array([[9.6820000e+03, 3.5105300e+05, 9.1357000e+04],
       [3.5105300e+05, 1.4399307e+07, 3.7720360e+06],
       [9.1357000e+04, 3.7720360e+06, 9.9835800e+05]])

**5. Compute the inverse of `XTX`.**

In [34]:
IXTX = np.linalg.inv(XTX)

In [35]:
IXTX

array([[ 9.19403586e-04, -3.66412216e-05,  5.43072261e-05],
       [-3.66412216e-05,  8.23303633e-06, -2.77534485e-05],
       [ 5.43072261e-05, -2.77534485e-05,  1.00891325e-04]])

**6. Create an array `y` with values `[950, 1300, 800, 1000, 1300]`.**

In [32]:
y = np.array([950, 1300, 800, 1000, 1300])

**7. Multiply the inverse of `XTX` with the transpose of `X`, and then multiply the result by `y`. Call the result `w`.**

In [40]:
w = IXTX.dot(X.T).dot(y)

In [41]:
w

array([23.12330961, -1.48124183,  5.69922946])

**8. What's the value of the last element of `w`?**

In [42]:
w[-1]

5.699229455065586